### DISCUSSION TITLE
Is this a safe way to use ulid concurrently?

### DISCUSSION THREAD

**Discussion by frederikhors** [ent/ent#2641] [Q&A]
_I just asked [this question on SO](https://stackoverflow.com/questions/72584598/is-this-a-safe-way-to-use-ulid-concurrently-with-other-libraries-too). I'm posting this here too both for me and for for future "researchers"._

I'm trying to use for the first time the [ulid package](https://github.com/oklog/ulid).

In [their README](https://github.com/oklog/ulid#usage) they say:

> Please note that `rand.Rand` from the `math` package is *not* safe for concurrent use.
> Instantiate one per long living go-routine or use a `sync.Pool` if you want to avoid the potential contention of a locked `rand.Source` as its been frequently observed in the package level functions.

Can you help me understand what does this mean and **how to write SAFE code for concurrent use with libraries such [ent](https://github.com/ent/ent) or [gqlgen](https://github.com/99designs/gqlgen)**?

Example: I'm using the below code in my app to generate new IDs (sometimes even many of them in the same millisecond which is fine for `ulid`).

```go
import (
  "math/rand"
  "time"

  "github.com/oklog/ulid/v2"
)

var defaultEntropySource *ulid.MonotonicEntropy

func init() {
  defaultEntropySource = ulid.Monotonic(rand.New(rand.NewSource(time.Now().UnixNano())), 0)
}

func NewID() string {
  return ulid.MustNew(ulid.Timestamp(time.Now()), defaultEntropySource).String()
}
```

**Is this a safe way to use the package?**

Can you help me better understand?

**Comment by a8m** [comment]
I think you can use something like [LockedSource](https://github.com/golang/exp/blob/bcd21879b8fdbb26ead5d311780450a7df02c3a2/rand/rand.go#L340).

Not near a computer atm and can’t test it. I hope it helps.

**Reply by a8m** [reply]
The global one is synced btw. You just need to set its seed. See: https://github.com/golang/go/blob/19309779ac5e2f5a2fd3cbb34421dafb2855ac21/src/math/rand/rand.go#L293

**Reply by frederikhors** [reply]
> You just need to set its seed

Yeah. But this is what I'm asking, how to?

And am I right saying that in the contrib/pulid example the code is not safe for concurrent use right now?

**Reply by a8m** [reply]
Why not? It uses the standard [crypto/rand.Reader](https://github.com/ent/contrib/blob/master/entgql/internal/todopulid/ent/schema/pulid/pulid.go#L20).

**Reply by frederikhors** [reply]
[`crypto/rand.Reader`](https://github.com/ent/contrib/blob/master/entgql/internal/todopulid/ent/schema/pulid/pulid.go#L20) is safe and `math.Rand` not?

[I'm scratching my head right now.]

**Reply by a8m** [reply]
The implementation is different, see examples:
https://github.com/golang/go/blob/master/src/crypto/rand/rand_unix.go#L34
https://github.com/golang/go/blob/master/src/crypto/rand/rand_plan9.go#L29

**Reply by frederikhors** [reply]
Running [this code](https://go.dev/play/p/CkdK1K_H_Wt) with `go run -race` I'm having many data race errors. So I think something is wrong with `crypto/rand` used like this.

I'm still studying and learning...

**Reply by frederikhors** [reply]
@tsenart is helping me a lot here: https://github.com/oklog/ulid/issues/79. He's also opening PRs for safe concurrent use for ulid.

**Reply by frederikhors** [reply]
> The implementation is different, see examples: https://github.com/golang/go/blob/master/src/crypto/rand/rand_unix.go#L34 https://github.com/golang/go/blob/master/src/crypto/rand/rand_plan9.go#L29

I do not understand what you mean. I think there is some issue with that pulid example. Maybe @giautm can help here.

**Reply by frederikhors** [reply]
The below is a simple code that doesn't work concurrently even with `crypto/rand`. And I'd like to know if and where I'm wrong.

```go
package main

import (
    "crypto/rand"
    "fmt"
    "log"
    "sync"
    "time"

    "github.com/oklog/ulid/v2"
)

func main() {
  var wg sync.WaitGroup

  entropy := ulid.Monotonic(rand.Reader, 0)

  ids := make([]string, 100)

  for i := 0; i < len(ids); i++ {
    wg.Add(1)

    go func(i int) {
      defer wg.Done()

      newID := ulid.MustNew(ulid.Timestamp(time.Now()), entropy)

      ids[i] = newID.String()
    }(i)
  }

  wg.Wait()

  fmt.Printf("now checking %v elements...", len(ids))

  seen := make(map[string]bool)

  for _, id := range ids {
    // fmt.Println(id)
    if seen[id] {
        fmt.Println(id)
        log.Fatal("duplicate found")
    }
    seen[id] = true
  }
}
```

**Reply by a8m** [reply]
Here’s the issue. The Source (and the Reader) are thread-safe, the MonotonicEntropy is not. See https://github.com/oklog/ulid/blob/master/ulid.go#L523

The PR you mentioned suggests how to solve this. We can add a comment near the example code in ent/contrib to help others in the future.

**Reply by a8m** [reply]
> And am I right saying that in the contrib/pulid example the code is not safe for concurrent use right now?

You’re right. Please, feel free to add a comment on the implementation 🙏🏻

**Reply by frederikhors** [reply]
Yeah. For now I have suspended the introduction of ulid in the project until I understand well how to do it. In the next few days I will also fix here with a PR.

**Reply by frederikhors** [reply]
The other problem is that there are so many ways to solve this; and there are also the number of processors to consider. It is explained VERY WELL here: https://qqq.ninja/blog/post/fast-threadsafe-randomness-in-go/.

I wish I was good enough to integrate those tips into the code. For example, what do you think of [the latest (faster) method he explains](https://qqq.ninja/blog/post/fast-threadsafe-randomness-in-go/#using-hashmaphash)? It would be easy to use it for our entropy?
